File: Data Structures & Algorithms/reorder-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        ListNode l = head;  
        int n = 0;
        while(l != null){
            l = l.next;
            n++;
        }
        n /= 2;
        l = head;
        while(l!=null || n > 0){
            ListNode r = l.next;
            while(r!=null && r.next != null && r.next.next!=null) r = r.next;
            if(r == null || r.next == null) return;
            r.next.next = l.next;
            l.next = r.next;
            r.next = null;
            l = l.next.next;
            n--;
        }
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        val pQ = PriorityQueue<ListNode>(compareBy{it.`val`})
        for(head in lists){
            var pom = head
            while(pom != null){
                pQ.add(pom)
                pom = pom.next
            }
        }
        val head = pQ.poll() ?: return null
        var pom = head
        while(pQ.isNotEmpty()){
            pom.next = pQ.poll()
            pom = pom.next!!
        }
        return head
    }
}
